core/shared/src/main/scala/fs2/Pull.scala

```diff
@@ -65,46 +74,72 @@ final class Pull[+F[_], +O, +R] private[fs2] (private[fs2] val free: FreeC[F, O,
   /** Lifts this pull to the specified resource type. */
   def covaryResource[R2 >: R]: Pull[F, O, R2] = this
 
+  def append[F2[x] >: F[x], O2 >: O, R2](post: => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
+    new Bind[F2, O2, R, R2](this) {
+      def cont(r: Result[R]): Pull[F2, O2, R2] = r match {
+        case _: Value[_]       => post
+        case r: Interrupted[_] => r
+        case r: Fail           => r
+      }
+    }
+
   /** Applies the resource of this pull to `f` and returns the result in a new `Pull`. */
-  def map[R2](f: R => R2): Pull[F, O, R2] = new Pull(free.map(f))
+  def map[R2](f: R => R2): Pull[F, O, R2] =
+    new Bind[F, O, R, R2](this) {
+      def cont(e: Result[R]): Pull[F, O, R2] = Result.map(e)(f)
+    }
 
   /** Applies the outputs of this pull to `f` and returns the result in a new `Pull`. */
-  def mapOutput[O2](f: O => O2): Pull[F, O2, R] = new Pull(free.mapOutput(f))
+  def mapOutput[O2](f: O => O2): Pull[F, O2, R]
 
   /** Run `p2` after `this`, regardless of errors during `this`, then reraise any errors encountered during `this`. */
   def onComplete[F2[x] >: F[x], O2 >: O, R2 >: R](p2: => Pull[F2, O2, R2]): Pull[F2, O2, R2] =
-    handleErrorWith(e => p2 >> new Pull(Result.Fail(e))) >> p2
+    handleErrorWith(e => p2 >> Fail(e)) >> p2
 
   /** If `this` terminates with `Pull.raiseError(e)`, invoke `h(e)`. */
   def handleErrorWith[F2[x] >: F[x], O2 >: O, R2 >: R](
       h: Throwable => Pull[F2, O2, R2]
   ): Pull[F2, O2, R2] =
-    new Pull(free.handleErrorWith(e => h(e).free))
+    new Bind[F2, O2, R2, R2](this) {
+      def cont(e: Result[R2]): Pull[F2, O2, R2] = e match {
+        case Fail(e) =>
+          try h(e)
+          catch { case NonFatal(e) => Fail(e) }
+        case other => other
+      }
+    }
 
   /** Discards the result type of this pull. */
   def void: Pull[F, O, Unit] = as(())
```

Comment:
@mpilquist BTW, and off-main topic: Should we remove the definitions of `Pull.as` and `Pull.void`, since they are given _for free_ by the Functor syntax of `cats`?